### PR TITLE
deployed-labeler: improve error handling, add GET endpoint

### PR DESCRIPTION
**Pull request by adrienthebo**
## Description

deployed-labeler: bump commit limit to 5 pages, 100 results

- - -

deployed-labeler: emit error messages from /deployed endpoint
    
This commit generates user readable error messages when the /deployed endpoint is called without required parameters.

- - -

deployed-labeler: expose log level CLI option, adjust log levels

- - -

deployed-labeler: add GET handler for deployed to indicate label status

- - -

deployed-labeler: serialise errors as strings
    
per https://github.com/golang/go/issues/5161 serializing an `error` object to json emits an empty object; this caused deployed-labeler to emit an empty object when erroring. This commit converts all errors to strings before returning to the user to provide more context.

## How to test

Test error handling
1. Create a malformed GitHub key
2. Run `curl "http://localhost:8080/deployed?commit=470062367333b5d9992188dbc9ff89465c0ab1e3&team=workspace"`
3. verify that an error is emitted and returned to the user.

Test status endpoint
1. Remove the deployed labels from https://github.com/gitpod-io/gitpod/pull/14215
2. run `curl "http://localhost:8080/deployed?commit=470062367333b5d9992188dbc9ff89465c0ab1e3&team=workspace"`
3. confirm that the correct PR is listed as unlabeled

## Release Notes

```release-note
```

### CODE REVIEW

**Comment by roboquat** [comment]
@adrienthebo: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: plugins/deployed-labeler/main.go [hunk 1]
```diff
@@ -90,27 +90,44 @@ func main() {
...
-			break
+	}
+	if team == "" || commitSHA == "" {
+
+		var preconditionFailed struct {
+			Errors []string
 		}
 
+		preconditionFailed.Errors = make([]string, 0, 1)
+		preconditionFailed.Errors = append(preconditionFailed.Errors, fmt.Sprintf("team and commit are required parameters. Team: '%v', commit: '%v'", team, commitSHA))
```
Comment (ArthurSens) [inline plugins/deployed-labeler/main.go:117]: What was your goal in creating a new type to store the error?

Could we use standard error libs? Maybe [errors](https://pkg.go.dev/errors) would make more sense here

File: plugins/deployed-labeler/main.go [hunk 2]
```diff
@@ -169,7 +171,7 @@ func (s *server) updatePullRequests(prs []pullRequest, team string) (teamDeploye
 
 		lblTeam := teamLabel(team)
 		if _, belongs := pr.Labels[lblTeam]; !belongs {
-			s.log.Infof("PR %v does not belong to %v, skipping it", pr.Number, team)
+			s.log.Debugf("PR %v does not belong to %v, skipping it", pr.Number, team)
```
Comment (ArthurSens) [inline plugins/deployed-labeler/main.go:221]: Why set all those log lines to debug? 😅

File: plugins/deployed-labeler/main.go [hunk 3]
```diff
@@ -90,36 +92,91 @@ func main() {
...
+	}
+
+	if team == "" || commitSHA == "" {
+
+		var preconditionFailed struct {
+			Errors []string
+		}
+
+		preconditionFailed.Errors = make([]string, 0, 1)
+		preconditionFailed.Errors = append(preconditionFailed.Errors, fmt.Sprintf("team and commit are required parameters. Team: '%v', commit: '%v'", team, commitSHA))
```
Comment (ArthurSens) [inline plugins/deployed-labeler/main.go:117]: What was your goal in creating a new type to handle errors?

Would it make sense to use go standard libraries? Example: the package [errors](https://pkg.go.dev/errors)

File: plugins/deployed-labeler/main.go [hunk 4]
```diff
@@ -140,6 +197,15 @@ func (s *server) handleMarkDeployedPRs(ctx context.Context, commitSHA, team stri
 	return s.updatePullRequests(prs, team)
 }
 
+func (s *server) handleGetUnmarkedPRs(ctx context.Context, commitSHA, team string) (labeledPRs, unlabeledPRs []string, errors []error) {
+	prs, err := s.getMergedPRs(ctx, commitSHA)
+	if err != nil {
+		return nil, nil, []error{err}
+	}
```
Comment (ArthurSens) [inline plugins/deployed-labeler/main.go:204]:
```suggestion
func (s *server) handleGetUnmarkedPRs(ctx context.Context, commitSHA, team string) (labeledPRs, unlabeledPRs []string, errors error) {
	prs, err := s.getMergedPRs(ctx, commitSHA)
	if err != nil {
		return nil, nil, err
	}
```

Why use an array if you'll never return more than one error?

**Review by ArthurSens — CHANGES_REQUESTED** [review]
Really nice addition of the `GET` endpoint! I can see that helping us a lot when debugging failures 🙂 

I've made just a few suggestions about go

**Comment by adrienthebo** [comment]
@ArthurSens thanks for the detailed review! This PR was the result of a bunch of fairly blind hacking and stumbling around Go's error handling; a lot of decisions were based on "well, it compiles, so good enough!" I'll run through and address your comments and do another round of general cleanup and will follow up with you on another review - likely on Monday since it's Friday night your time, but perhaps tonight if you're burning the weekend oil. 😊 No rush!

**Comment by kylos101** [comment]
Hi @gitpod-io/engineering-delivery-operations-experience it looks like this is caught in limbo, so I added you all back as a reviewer.

**Comment by liam-j-bennett** [comment]
@kylos101 @adrienthebo This work also might have gotten caught up in the changes in the team structure (i.e. Scout Team). Could this work be handed off?

**Comment by kylos101** [comment]
> @kylos101 @adrienthebo This work also might have gotten caught up in the changes in the team structure (i.e. Scout Team). Could this work be handed off?

Hey @liam-j-bennett :wave:  we discussed with @meysholdt and @corneliusludmann during the platform/self-hosted to devx/delivery transformation, and my understanding is that Delivery Team owns the labeling component. 

Is it in the list of Delivery owned components? Perhaps something changed since then? cc: @corneliusludmann

**Comment by corneliusludmann** [comment]
> Is it in the list of Delivery owned components? Perhaps something changed since then?

No, nothing has changed. I think @liam-j-bennett's comment was about Adrien being busy with scout team work and not in the position to continue this and should be handed over to someone else from the delivery team.

I removed Adrien from the corresponding issue https://github.com/gitpod-io/gitbot/pull/86#issuecomment-1335248996 and put it back to “Scheduled” so that someone else can pick this up soon.

File: plugins/deployed-labeler/main.go (hunk 4 above)
Comment (Pothulapati) [reply plugins/deployed-labeler/main.go:204]: > Why use an array if you'll never return more than one error?

This seems to have done as `s.findTeamPullRequests` has a return type of `(labeled, unlabeled []string, errs []error)`. We are just passing those over for the caller to use as needed.

File: plugins/deployed-labeler/main.go (hunk 2 above)
Comment (Pothulapati) [reply plugins/deployed-labeler/main.go]: Moved them back as this info seems useful

File: plugins/deployed-labeler/main.go (hunk 1 above)
Comment (Pothulapati) [reply plugins/deployed-labeler/main.go]: > What was your goal in creating a new type to store the error?

The PR mention's _per https://github.com/golang/go/issues/5161 serializing an error object to json emits an empty object; this caused deployed-labeler to emit an empty object when erroring. This commit converts all errors to strings before returning to the user to provide more context._

Seems to be serialisation and making sure the errors are returned in the HTTP response. 🤔

**Comment by Pothulapati** [comment]
Now that the PR compiles, Testing this PR now. 👀

**Comment by Pothulapati** [comment]
Hmm. Like @ArthurSens suggested, There seems to be no requirement for a new `err` type for serialising them into the json object. It already seems to work.

```
gitpod /workspace/gitbot $ curl -XPOST "http://localhost:8080/deployed?commit=01f4897c5323433e7831ca948f7d340c3c762&team=webapp"
{
  "deployedPRs": {
    "team": null,
    "all": null
  },
  "errors": [
    [
      {
        "Message": "Variable $commit of type GitObjectID! was provided invalid value",
        "Locations": [
          {
            "Line": 1,
            "Column": 7
          }
        ]
      }
    ]
  ]
}

```

It does not  work in the new precondition thing though (because of the lack of type), and I've reduced it to return a map with `err` field instead. Making the following work:

```
gitpod /workspace/gitbot $ curl -XPOST "http://localhost:8080/deployed?commit=01f4897c5323433e7831ca948f7d340c3c762&team="
{
  "error": "team and commit are required parameters"
}
```

File: plugins/deployed-labeler/main.go (hunk 1 above)
Comment (Pothulapati) [reply plugins/deployed-labeler/main.go]: Update! New err type is not required. Rolled it back to `error`

**Comment by Pothulapati** [comment]
Update, Now that the changes work:

The move from `Error` to `String` for returning errors seem to be necessary to handle Internal API errors that Github Client returns. This is useful when there is a malfunctioning token that is passed, etc.

```
{
  "deployedPRs": {
    "team": null,
    "all": null
  },
  "errors": [
    "non-200 OK status code: 401 Unauthorized body: \"{\\\"message\\\":\\\"Bad credentials\\\",\\\"documentation_url\\\":\\\"https://docs.github.com/graphql\\\"}\""
  ]
}
```

File: plugins/deployed-labeler/main.go (hunk 4 above)
Comment (Pothulapati) [reply plugins/deployed-labeler/main.go]: Right. Updated it to use a single error.

**Comment by Pothulapati** [comment]
PR is now ready to review! Cleaned up commit messages by rebasing things. 👍🏼

**Comment by Pothulapati** [comment]
@ArthurSens Just to confirm before proceeding, The deployment process is manual as per the[readme doc](https://github.com/gitpod-io/gitbot/tree/main/plugins/deployed-labeler#deploying) right? :)

**Comment by ArthurSens** [comment]
> @ArthurSens Just to confirm before proceeding, The deployment process is manual as per the[readme doc](https://github.com/gitpod-io/gitbot/tree/main/plugins/deployed-labeler#deploying) right? :)

It has been a long time since I did anything here, but I believe it still the same process 😅

**Pull request merged**